### DISCUSSION TITLE
ESP8266-12F, 8-Channel Relay and Tasmota

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#12473] [Q&A] [answered]
Hello, 

I was trying to Max out my custom device running on ESP8266-12F Chip. I was a bit apprehensive as I had read that most GIPO-s on this chip are problematic and only a handful can be freely used. In my setup, using Tasmota's "Generic Template" I used 13 of the 14 available GPIO-s in the template and So far the device seems to be doing alright without any signs of trouble. Seems to me that "Tasmota" firmware is so well written that it has overcome the problems that many programmers face when making firmware for ESP8266 at least when it is for a simple device to control relays using switches or buttons. (Though on startup/boot there is small hiccup in form of momentary flicker of attached lights on some relays but I think that is acceptable.)

In my setup I am using 8 Relays, 4 Switches, and 1 button. 
(EDIT: I am using 700ma 5v power supply that gives dual output of 5v and 3.3 volts using AMS1117 3.3 chip. It seems to be sufficient to power 3.3v ESP8266 chip and 8 channel 5v relay.)

My question is, since I am using GPIO Numbers 9,10,15, and 16 am I setting myself up for an unreliable device down the road? Lets say the ESP8266, for what ever reason, one day decides to contact the manufacturer automatically with diagnostic information, will it crash, freeze or have other undesirable symptoms since all the GPIOS are controlled by Tasmota? or am I safe to assume that with my Tasmota configuration I am in safe hands as long as there is no hardware failure in the Power Supply, PCB and the Relay? Below is the picture of my fully populated template:

![Screenshot (204)](https://user-images.githubusercontent.com/69470381/123375340-39721f80-d5a6-11eb-85bc-a47a990cdbba.png)

Edit: Here is the 8Channel Beast:

![8CH relay](https://user-images.githubusercontent.com/69470381/123380933-af2db980-d5ad-11eb-9c2f-50dc280329f4.jpeg)

Regards,

Navi

**Comment by barbudor** [comment, marked as answer]
> running on ESP8266-12F
>  I am using GPIO Numbers 9,10,15, and 16

You shouldn't be using GOIO9 and 10 on an ESP8266. If you fool Tasmota by setting SO51, you just taking the risk of crashing it
Here is the schematic of ESP12F where you can see GPIO9 and 10 are connected to the flash chip.
Just don't do it

![image](https://user-images.githubusercontent.com/2996042/123379781-e3de4880-d58e-11eb-9b14-aaa6e2ffc9fd.png)

15 is not a problem as a relay output. At reset pin is hi-Z and there must be a pull-down on your board so your relay will be always ON during the boot/reboot of the the ESP until Tasmota starts.

0 and 2 must have a pull-up to insure proper boot, so the relay will always be OFF during boot/reboot.
Of course, pray for that nobody press that Button on PIO2 when the device is rebooting

1 is TXD that must nt be driven low at boot. AS you are driving a relay_i you must have a pull-up here which will makes your relay OFF during boot.

>  Lets say the ESP8266, for what ever reason, one day decides to contact the manufacturer automatically with diagnostic information

It is well know that the ESP is spying for any information that is against the Chinese regime. If the ESP detect such activity it will report to the Chinese government which in turn will requires Huawei to shut down all 5G relays in your neighborhood.


So in clear, your setup is not safe.
Consider ESP32 or use an I2C extender.

**Reply by NsinghP** [reply]
> It is well know that the ESP is spying for any information that is against the Chinese regime. If the ESP detect such activity it will report to the Chinese government which in turn will requires Huawei to shut down all 5G relays in your neighborhood.

LOL, You must be joking, right?

**Reply by NsinghP** [reply]
> You shouldn't be using GOIO9 and 10 on an ESP8266. If you fool Tasmota by setting SO51, you just taking the risk of crashing it
> Here is the schematic of ESP12F where you can see GPIO9 and 10 are connected to the flash chip.
> Just don't do it

Thanks, I will keep an eye out for strange behavior. So far both these pins along with other GPIOs are working great. EDIT: All I did was solder the wires and configured the Template, everything else seems to be working like a charm.

**Reply by barbudor** [reply]
> LOL, You must be joking, right?

Your guess is as good as mine 🤣🤣🤣

**Reply by NsinghP** [reply]
> Your guess is as good as mine 🤣🤣🤣

Look what you've done, now I will have to dig deeper until I figure it out! :)

**Reply by sfromis** [reply]
If you want to use gpio 9/10, get a module with ESP8285, where the pins are actually available for general use.

**Reply by NsinghP** [reply]
> get a module with ESP8285

Thanks, 8285 modules that are available online only have 11 GPIOS broken out and 9/10 are missing.  

Based on what I read in another forum I think the reason I am not experiencing any problems with GPIO-9 and 10 is: 

(a)  maybe because I used DIO mode to do erase/flash of the fresh chip the first time, I am assuming, when I flashed with Tasmotizer it automatically meant DIO mode of flashing. 

(b) I was also able to do OTA upgrade of this device without any problems. So I think I am not having any firmware flashing issues either. (unless OTA flashing is significantly different than wire flashing) But even then I think I will not have any flashing issues because during wired flashing I disconnect all peripherals.  

The only way I think my setup can do damage is by somehow destroying the flash chip so that it is impossible to write to it again.

I have my fingers crossed...

**Reply by Jason2866** [reply]
Tasmota is using flash mode DOUT as standard for **all** ESP82XX not DIO.
(Tasmotizer uses DOUT) If this is the case you haved changed without a reason. DOUT works for all devices, all other flash modes do not. It seems you do not know what are the differences of the flash modes.
Using GPIO 9 10 is a very bad idea with a ESP8266. The are connected to the flash chip.
Your setup is against the design recommendations from espressif and flash chip manufactors.
Probably you will damage either the ESP8266 or the flash or both.

In short. Do NOT do this.

**Reply by NsinghP** [reply]
> Tasmota is using flash mode DOUT as standard for **all** ESP82XX not DIO.
...

Thanks for correcting my wrong assumption.

> Using GPIO 9 10 is a very bad idea with a ESP8266. The are connected to the flash chip.
... 

Agreed, that's what most of ESP literature online seems to indicate.

> Your setup is against the design recommendations from espressif and flash chip manufactors.
> Probably you will damage either the ESP8266 or the flash or both.

I hope I am able to understand what happened when it finally starts to act up or completely gets fried.

**Reply by sfromis** [reply]
You can always get at ESP8285 gpio 9/10 soldering directly to the chip.
More practical may be the easier access on (at least) the PSF-A85 module.

**Reply by NsinghP** [reply]
> You can always get at ESP8285 gpio 9/10 soldering directly to the chip.
> More practical may be the easier access on (at least) the PSF-A85 module.

Thanks, I will try to get both of them and test it out. But one limitation I see is that ESP8285 only has 1MB FLash. That will make it hard to load all tasmota features I want, especially MQTT-TLS and File management.

**Reply by sfromis** [reply]
Easy path to more pins is ESP32, where Tasmota also runs.

**Reply by NsinghP** [reply]
> Easy path to more pins is ESP32, where Tasmota also runs.

Yes, I tried that, had super hard time soldering it, so gave up. But will revisit it as I am getting comfortable with tight soldering.

**Reply by NsinghP** [reply]
>  If you fool Tasmota by setting SO51, you just taking the risk of crashing it...

Looked up SO51, No I am not using it;

14:56:31.907 CMD: setoption51
14:56:31.912 RSL: RESULT = {"SetOption51":"OFF"}

**Reply by sfromis** [reply]
That makes no difference to the badness of what you're doing. SetOption51 *only* affects whether the pins are included in "Configure Module". If you use gpio 9/10 (in this case via a template), the option is simply not used.

**Reply by NsinghP** [reply]
I google searched and found an old thread titled "GPIO pin 9 mystery", this maybe an explanation of why in my setup GPIO 9 and 10 are working flawlessly. Maybe I just lucked out :) I also read up on DOUT and DIO mode and it seems they both use two SPI pins, that may also be a factor.

"-------Start Quote--------
#1 Post by Martinus » 17 Jun 2017, 19:18

So it seems that pin GPIO-9 (D11) can not be used on some boards, leading to WDT errors. I've checked this today on my NodeMCU board and all GPIO and even PWM commands on port gpio-9 as well as gpio-10 run without issues.
I got these boards at the time that they were just released and could not even program these properly because the IDE was still using QIO mode. Once changed to DIO mode they started to work...

So it looks like the ESP modules on my NodeMCU board have the flash chip connected using two data lines only and GPIO-9 and GPIO-10 are not connected to the flash chip. So the NodeMCU board can use them as free gpio pins.

But it also seems that some (or a lot?) are using ESP modules that have the flash chip connected with 4 datalines (QIO). Then you are out of luck. Maybe these ESP modules look like the newer ones with extra pins on the bottom side but they are likely not wired in the same way as mine.

How to tell? I guess only by taking the shield off and examine pcb traces on the ESP module...

But am i the only one that can use GPIO-9 without errors? Feedback required here. Maybe we can write a wiki on this subject once we figure out which boards could use GPIO-9.


---------End Quote----------"

Edit: FYI: Here is a picture of two ESP8266Chips I have. One on the right is a chip I destroyed while practicing to solder it, this chip says ("QIO L4). Chip on the left with no "QIO" written on it is the one I have recently flashed and is working without problems with GPIO-9 and 10. Unfortunately chip on right is destroyed or I would have confirmed if it works with GPIO-9 and 10 or not. So it seems depending on the ESP8266-12F manufacturer, some chips may work very well with restricted GPIO-9/10. But there is no sure shot way of knowing which chip will work or not on boards that have the chip pre-soldered as "QIO" identification mark gets hidden on soldered side .

![ESP8266 compared](https://user-images.githubusercontent.com/69470381/123427468-4b71b380-d5e2-11eb-9a0a-7c7581a75ee9.jpeg)

**Reply by Jason2866** [reply]
You can remove the metall shield and lift the pin3 and pin7 from the flash chip.
So GPIO 9 / 10 can be used when using flash mode DOUT (Tasmota standard)
Is this really worth the effort? Imho no.
Try to source a ESP-12 module with a ESP8285 2M variant inside

**Reply by NsinghP** [reply]
> 
> 
> You can remove the metall shield and lift the pin3 and pin7 from the flash chip.
> So GPIO 9 / 10 can be used when using flash mode DOUT (Tasmota standard)
> Is this really worth the effort? Imho no.
> Try to source a ESP-12 module with a ESP8285 2M variant inside

Right, it is not worth the effort. But for someone using the bare chip, it may be relatively easy to identify which chip to use if they need more GPIOs, that is if every manufacturer marks and identifies their QIO version. I'm not sure if that is standard practice though. In my case, since I'm using bare chip that does not have the mark, I am now somewhat comfortable using my 8-Relay version. I hope it keeps working. 

Thanks for your time, 

Navi

**Reply by sfromis** [reply]
You are not using a "bare chip", but a whole module including two chips and some other components. And I'd not bank on details about internals always being represented on silk screened labeling.

**Reply by NsinghP** [reply]
> You are not using a "bare chip", but a whole module including two chips and some other components.

Yes, thanks for pointing that out, I too realized that when I peeled the metal casing from one of the chips.

**Reply by NsinghP** [reply]
> ... And I'd not bank on details about internals always being represented on silk screened labeling.

Agreed, somehow I am getting a feeling that unlabeled one I am using is a counterfeit.  But very difficult to figure out what is genuine and what is fake without the experience of handling thousands upon thousands of these chips.

**Reply by sfromis** [reply]
That's easy. Those modules are all clones, with variations. Nothing like "genuine" or not.

**Comment by lalo-uy** [comment]
You can get shield less modules, like this one:

US $1.72  9％ Off | NodeMCU V3 Lua WIFI module integration of ESP8266 +
extra memory 32M Flash, USB-serial CH340G
https://star.aliexpress.com/share/share.htm?image=Uf069c82d6e47436ca3de1d6a6e271e65X.jpg&businessType=ProductDetail&title=US%20%241.72%20%209%EF%BC%85%20Off%20%7C%20NodeMCU%20V3%20Lua%20WIFI%20module%20integration%20of%20ESP8266%20%2B%20extra%20memory%2032M%20Flash%2C%20USB-serial%20CH340G&platform=AE&redirectUrl=https%3A%2F%2Fwww.aliexpress.com%2Fitem%2F1005002006567767.html%3F%26srcSns%3Dsns_Copy%26tid%3Dwhite_backgroup_101%26mb%3D63ADwDKl9kzJNpO%26businessType%3DProductDetail%26spreadType%3DsocialShare

**Comment by lalo-uy** [comment]
If a ESP8285 is ok, take a look at ESP-01M modules.

**Reply by NsinghP** [reply]
Thanks for your suggestions, since I had ordered few custom pcbs, I am afraid that I am stuck with using esp8266 12f.  And after this I am planning on experimenting with esp32.
